scalapact-core/src/main/scala/com/itv/scalapactcore/common/matching/InteractionMatchingPrograms.scala

```diff
@@ -0,0 +1,213 @@
+  val success: Boolean = {
+    (method, path, headers, body) match {
+      case (MatchMethodSuccess, MatchPathSuccess, MatchHeadersSuccess, MatchBodySuccess) => true
+      case _ => false
+    }
+  }
+
+  val failed: Boolean = !success
+}
+
+case class ResponseMatchOutcome(status: MatchStatusOutcome, headers: MatchHeadersOutcome, body: MatchBodyOutcome) {
+  val success: Boolean = {
+    (status, headers, body) match {
+      case (MatchStatusSuccess, MatchHeadersSuccess, MatchBodySuccess) => true
+      case _ => false
+    }
+  }
+
+  val failed: Boolean = !success
+}
+
+object InteractionMatchingPrograms {
+
+  type MatchResponseStage[A] = Free[MatchResponseStageA, A]
+
+  // Response
+  private def matchResponseStatus(expected: InteractionResponse, actual: InteractionResponse): MatchResponseStage[MatchStatusOutcome] =
+    Free.liftF(MatchResponseStatus(expected, actual))
+
+  private def matchResponseHeaders(expected: InteractionResponse, actual: InteractionResponse): MatchResponseStage[MatchHeadersOutcome] =
+    Free.liftF(MatchResponseHeaders(expected, actual))
+
+  private def matchResponseBody(expected: InteractionResponse, actual: InteractionResponse): MatchResponseStage[MatchBodyOutcome] =
+    Free.liftF(MatchResponseBody(expected, actual))
+
+  def matchResponseProgram(expected: InteractionResponse, actual: InteractionResponse): MatchResponseStage[ResponseMatchOutcome] =
+    for {
+      s <- matchResponseStatus(expected, actual)
+      h <- matchResponseHeaders(expected, actual)
+      b <- matchResponseBody(expected, actual)
+    } yield ResponseMatchOutcome(s, h, b)
+
+  type MatchRequestStage[A] = Free[MatchRequestStageA, A]
+
+  // Request
+  private def matchRequestMethod(expected: InteractionRequest, actual: InteractionRequest): MatchRequestStage[MatchMethodOutcome] =
+    Free.liftF(MatchRequestMethod(expected, actual))
+
+  private def matchRequestPath(expected: InteractionRequest, actual: InteractionRequest): MatchRequestStage[MatchPathOutcome] =
+    Free.liftF(MatchRequestPath(expected, actual))
+
+  private def matchRequestHeaders(expected: InteractionRequest, actual: InteractionRequest): MatchRequestStage[MatchHeadersOutcome] =
+    Free.liftF(MatchRequestHeaders(expected, actual))
+
+  private def matchRequestBody(expected: InteractionRequest, actual: InteractionRequest): MatchRequestStage[MatchBodyOutcome] =
+    Free.liftF(MatchRequestBody(expected, actual))
+
+  def matchRequestProgram(expected: InteractionRequest, actual: InteractionRequest): MatchRequestStage[RequestMatchOutcome] =
+    for {
+      m <- matchRequestMethod(expected, actual)
+      p <- matchRequestPath(expected, actual)
+      h <- matchRequestHeaders(expected, actual)
+      b <- matchRequestBody(expected, actual)
+    } yield RequestMatchOutcome(m, p, h, b)
+
+}
+
+object MatchingInterpreters {
+
+  object Request {
+
+    object permissive extends (MatchRequestStageA ~> Id) {
```

Comment:
`~>` is the Natural Transformation from one type of Monad to another in this case from Free to Id. I'm using `Id` because I don't need anything more sophisticated.
If it makes it easier to see how this works, it could be:
`object permissive extends NaturalTransformation[F, K]` it's just shorthand for that.